elastic/beats
# metricbeat/module/redis: Add TLS and username support for Redis module

## What does this PR do?

PR adds support for TLS to Redis module in metricbeat. It also adds supports for accepting username (supported by by Redis 6.0+) from URI's userinfo or query parameter.

It has also has some nitpick changes pointed out by linters, etc.



## Checklist


## Author's Checklist

- TLS support
- Username support
- Testing with Redis 5.x or lower
- Testing with Redis 6.x or higher

## How to test this PR locally

- Use a redis-server with TLS enabled
- Use a redis-server with username set (Redis 6.0+ should be used)
- Use a redis-server with password set and without username
- Use a redis-server with both TLS and AUTH (username and password set) enabled.

But how will get the certificate, CA, keys, etc.? 

- Invoke the script: https://github.com/redis/redis/blob/unstable/utils/gen-test-certs.sh (after cloning redis/redis)
- It'll generate the following:
  - tests/tls/ca.{crt,key}          Self signed CA certificate.
  - tests/tls/redis.{crt,key}       A certificate with no key usage/policy restrictions.
  - tests/tls/client.{crt,key}      A certificate restricted for SSL client usage.
  - tests/tls/server.{crt,key}      A certificate restricted for SSL server usage.
  - tests/tls/redis.dh              DH Params file.
- Now run the redis-server in TLS mode (example):
```
$ redis-server --tls-port 6378 --port 0 --tls-cert-file ./tests/tls/redis.crt --tls-key-file ./tests/tls/redis.key --tls-ca-cert-file ./tests/tls/ca.crt
```
- If you want to use redis-cli or any other client (metricbeat's redis module) then use the same `ca.crt`, `redis.crt` and `redis.key`:
```
redis-cli -p 6378 --tls --cert ./tests/tls/redis.crt --key ./tests/tls/redis.key --cacert ./tests/tls/ca.crt
```

## Related issues

- Fixes https://github.com/elastic/integrations/issues/5427
